{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sys\n",
    "import warnings\n",
    "sys.path.insert(0, '/data16/marcus/scripts/nanoporePipelineScripts')\n",
    "import nanoporePipelineCommon as npCommon\n",
    "\n",
    "from tqdm.notebook import tqdm\n",
    "\n",
    "import seaborn as sea\n",
    "import matplotlib.pyplot as plt\n",
    "import plotly.express as px\n",
    "import plotly.graph_objects as go\n",
    "import plotly.io as pio\n",
    "pio.renderers.default = \"browser\"\n",
    "\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import statistics as stats\n",
    "pd.set_option('display.width', 200)\n",
    "pd.set_option('display.max_columns', None)\n",
    "\n",
    "CONVERSION_DICT = {\"xrn-1-5tera\": \"oldN2\",\n",
    "                   \"xrn-1-5tera-smg-6\": \"oldS6\",\n",
    "                   \"5tera_xrn-1-KD_wt\": \"newN2\",\n",
    "                   \"5tera_xrn-1-KD_smg-5\": \"newS5\",\n",
    "                   \"5tera_xrn-1-KD_smg-6\": \"newS6\",\n",
    "                   \"5tera_xrn-1-KD_smg-7\": \"newS7\",\n",
    "                   \"5tera_xrn-1-KD_wt_rerun\": \"newerN2\",\n",
    "                   \"5tera_xrn-1-KD_smg-6_rerun\": \"newerS6\",\n",
    "                   \"5tera_xrn-1-KD_smg-5_rerun\": \"newerS5\",\n",
    "                   \"sPM57\": \"sPM57\",\n",
    "                   \"sPM58\": \"sPM58\",\n",
    "                   }\n",
    "REV_CONVERSION_DICT = {val: key for key, val in CONVERSION_DICT.items()}\n",
    "\n",
    "print(f\"Imports done at {npCommon.get_dt(for_print=True)}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Load pre-processed files:"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "regenerate = False\n",
    "libs_to_load = sorted({\n",
    "    'oldN2',\n",
    "    'newN2',\n",
    "    'newerN2',\n",
    "    'oldS6',\n",
    "    # 'newS6',\n",
    "    'newerS6',\n",
    "    # 'newS5',\n",
    "    'newerS5',\n",
    "    'newS7',\n",
    "})\n",
    "\n",
    "try:\n",
    "    if regenerate:\n",
    "        raise FileNotFoundError\n",
    "    \n",
    "    reads_df_raw_path = npCommon.find_newest_matching_file(f\"./output_files/mega_merge_parquets/*_{'-'.join(libs_to_load)}_merged5TERA.reads_df.parquet\")\n",
    "    compressed_df_genes_raw_path = npCommon.find_newest_matching_file(f\"./output_files/mega_merge_parquets/*_{'-'.join(libs_to_load)}_merged5TERA.compressed_df.parquet\")\n",
    "    print(f\"Found preprocessed files at:\\n\\t{reads_df_raw_path}\\nand:\\n\\t{compressed_df_genes_raw_path}\")\n",
    "\n",
    "    reads_df_genes_raw = pd.read_parquet(reads_df_raw_path)\n",
    "    compressed_df_genes_raw = pd.read_parquet(compressed_df_genes_raw_path)\n",
    "except FileNotFoundError:\n",
    "    print(f\"Could not find preprocessed files matching these libs: {'/'.join(libs_to_load)}\\nGoing to create new ones from scratch! This will take longer.\")\n",
    "    reads_df_genes_raw, compressed_df_genes_raw = npCommon.load_and_merge_lib_parquets([REV_CONVERSION_DICT[lib] for lib in libs_to_load],\n",
    "                                                                                       drop_sub_n=1,\n",
    "                                                                                       add_tail_groupings=False,\n",
    "                                                                                       drop_failed_polya=False,\n",
    "                                                                                       group_by_t5=True,\n",
    "                                                                                       use_josh_assignment=False)\n",
    "    print(f\"Saving new parquets to speed up future runs.\")\n",
    "    reads_df_genes_raw.to_parquet(f\"./output_files/mega_merge_parquets/{npCommon.get_dt()}_{'-'.join(libs_to_load)}_merged5TERA.reads_df.parquet\")\n",
    "    compressed_df_genes_raw.to_parquet(f\"./output_files/mega_merge_parquets/{npCommon.get_dt()}_{'-'.join(libs_to_load)}_merged5TERA.compressed_df.parquet\")\n",
    "print(f\"Lib load done @ {npCommon.get_dt(for_print=True)}\")\n",
    "\n",
    "compressed_df_genes_short = compressed_df_genes_raw.copy()[[\"lib\", \"chr_id\", \"gene_id\", \"gene_name\", \"t5\", \"gene_hits\", \"gene_rpm\"]]\n",
    "compressed_df_genes_short.query(\"gene_name == 'rpl-12'\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Format libraries as columns with shortened names"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "conversion_dict = CONVERSION_DICT\n",
    "ans = [y for x, y in compressed_df_genes_short.groupby(['lib', 't5'], as_index=False)]\n",
    "df_dict = {}\n",
    "for i, df in enumerate(ans):\n",
    "    lib = df.lib.unique()[0]\n",
    "    t5 = df.t5.unique()[0]\n",
    "    df = df[[\"chr_id\", \"gene_id\", \"gene_name\", \"gene_hits\", \"gene_rpm\"]]\n",
    "    df = df.rename(columns={col: f'{col}_{conversion_dict[lib]}_t5{t5}' for col in df.columns if col not in [\"chr_id\", \"gene_id\", \"gene_name\"]})\n",
    "    df_dict[(conversion_dict[lib], t5)] = df.set_index([\"chr_id\", \"gene_id\", \"gene_name\"])\n",
    "    # print((conversion_dict[lib], t5))\n",
    "    # print(df_dict[(conversion_dict[lib], t5)].query(\"gene_name == 'rpl-12'\"))\n",
    "\n",
    "super_df = pd.concat(df_dict.values(), axis=1, join='outer').fillna(0)\n",
    "super_df"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# This step will calculate total hits and the fraction adapted for each gene\n",
    "\n",
    "filter_df_raw = pd.DataFrame()\n",
    "for lib in libs_to_load:\n",
    "    for rpm_or_hits in [\"rpm\", \"hits\"]:\n",
    "        super_df[f\"total_gene_{rpm_or_hits}_{lib}\"] = super_df[f\"gene_{rpm_or_hits}_{lib}_t5+\"] + super_df[f\"gene_{rpm_or_hits}_{lib}_t5-\"]\n",
    "    super_df[f\"fraction_adapted_{lib}\"] = super_df[f\"gene_hits_{lib}_t5+\"] / super_df[f\"total_gene_hits_{lib}\"]\n",
    "    \n",
    "    cols_to_carry_over = [col for col in super_df.columns if lib in col]\n",
    "    filter_df_raw[cols_to_carry_over] = super_df[cols_to_carry_over]\n",
    "filter_df_raw"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from scipy.stats import chi2_contingency, chisquare, fisher_exact, boschloo_exact, barnard_exact\n",
    "def row_chi2(row, target_lib_1, target_lib_2):\n",
    "    array = np.array([[row[f\"gene_hits_{target_lib_1}_t5-\"], row[f\"gene_hits_{target_lib_2}_t5-\"]],\n",
    "                      [row[f\"gene_hits_{target_lib_1}_t5+\"], row[f\"gene_hits_{target_lib_2}_t5+\"]]])\n",
    "    try:\n",
    "        chi2, p, deg_of_free, expected = chi2_contingency(array)\n",
    "        return chi2, p\n",
    "    except ValueError:\n",
    "        return None, None\n",
    "\n",
    "def row_fishers_exact(row, target_lib_1, target_lib_2, hits_or_rpm='hits', alternative='two-sided'):\n",
    "    array = np.array([[row[f\"gene_{hits_or_rpm}_{target_lib_1}_t5-\"], row[f\"gene_{hits_or_rpm}_{target_lib_2}_t5-\"]],\n",
    "                      [row[f\"gene_{hits_or_rpm}_{target_lib_1}_t5+\"], row[f\"gene_{hits_or_rpm}_{target_lib_2}_t5+\"]]])\n",
    "    if alternative not in ['two-sided', 'greater', 'less']:\n",
    "        raise KeyError(f\"Please use 'two-sided', 'greater', or 'less' for the alternative hypothesis input for fisher's exact test!!\")\n",
    "    odds_ratio, p_value = fisher_exact(array, alternative=alternative)\n",
    "    return odds_ratio, p_value\n",
    "\n",
    "def row_boschloo_exact(row, target_lib_1, target_lib_2, hits_or_rpm='hits', alternative='two-sided', sampling_points=32):\n",
    "    array = np.array([[row[f\"gene_{hits_or_rpm}_{target_lib_1}_t5-\"], row[f\"gene_{hits_or_rpm}_{target_lib_2}_t5-\"]],\n",
    "                      [row[f\"gene_{hits_or_rpm}_{target_lib_1}_t5+\"], row[f\"gene_{hits_or_rpm}_{target_lib_2}_t5+\"]]])\n",
    "    if alternative not in ['two-sided', 'greater', 'less']:\n",
    "        raise KeyError(f\"Please use 'two-sided', 'greater', or 'less' for the alternative hypothesis input for Boschloo's exact test!!\")\n",
    "    boschloo_result = boschloo_exact(array, alternative=alternative, n=sampling_points)\n",
    "    return boschloo_result.statistic, boschloo_result.pvalue\n",
    "\n",
    "def row_barnard_exact(row, target_lib_1, target_lib_2, hits_or_rpm='hits', alternative='two-sided', sampling_points=32):\n",
    "    array = np.array([[row[f\"gene_{hits_or_rpm}_{target_lib_1}_t5-\"], row[f\"gene_{hits_or_rpm}_{target_lib_2}_t5-\"]],\n",
    "                      [row[f\"gene_{hits_or_rpm}_{target_lib_1}_t5+\"], row[f\"gene_{hits_or_rpm}_{target_lib_2}_t5+\"]]])\n",
    "    if alternative not in ['two-sided', 'greater', 'less']:\n",
    "        raise KeyError(f\"Please use 'two-sided', 'greater', or 'less' for the alternative hypothesis input for Barnard's exact test!!\")\n",
    "    barnard_result = barnard_exact(array, alternative=alternative, n=sampling_points)\n",
    "    return barnard_result.statistic, barnard_result.pvalue"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Use the Stat Tests!\n",
    "\n",
    "Realization on 6/22/2023:\n",
    "I think I should be doing the cutoff filtering for all libraries at once, not just the target libraries. I have been doing it per-lib because that makes more sense in the context of \"THAT TEST\" but now the paper is more using this as a comparison between newerS6 and newerS5 effects. So I should just stick to filtering all genes, then do the stats... Damn."
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "\n",
    "filter_df = filter_df_raw.copy()\n",
    "\n",
    "output_dir = f\"/home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python/{npCommon.get_dt()}_fishers\"\n",
    "Path(output_dir).mkdir(parents=True, exist_ok=True)\n",
    "\n",
    "stat_test_record_file = Path(f\"{output_dir}/{npCommon.get_dt()}_stat_test_record_file.txt\")\n",
    "\n",
    "stat_test_record_file.write_text(f\"{npCommon.get_dt(for_print=True)}\\n\")\n",
    "\n",
    "def print_and_write(string):\n",
    "    print(string)\n",
    "    with open(stat_test_record_file, 'a') as f:\n",
    "        f.write(string + \"\\n\")\n",
    "\n",
    "# p-value setpoint and the applied cutoffs will be used for a Bonferroni correction\n",
    "#   Currently no genes will be dropped based on these filters/cutoffs\n",
    "base_sig_cutoff = 0.05\n",
    "\n",
    "cumulative_min_read_cutoff = 100\n",
    "filter_with_rpm_or_hits = \"hits\"\n",
    "filter_col_target = \"total\"  # \"total\" or \"adapted\" or \"unadapted\"\n",
    "tests_to_run = [\n",
    "    \"fishers\",\n",
    "    # \"boschloo\",\n",
    "    # \"barnard\",\n",
    "    # \"chi2\",\n",
    "]\n",
    "\n",
    "p_value_cutoff_dict = {}\n",
    "\n",
    "lib_comparisons_to_test = []\n",
    "# lib_comparisons_to_test += list(zip([\"newN2\"]*3, [\"newS5\", \"newS6\", \"newS7\"]))\n",
    "# lib_comparisons_to_test += [(\"newN2\", \"newS6\")]\n",
    "lib_comparisons_to_test += [(\"oldN2\", \"oldS6\")]\n",
    "lib_comparisons_to_test += [(\"newerN2\", \"newerS6\")]\n",
    "lib_comparisons_to_test += [(\"newerN2\", \"newerS5\")]\n",
    "lib_comparisons_to_test += [(\"newN2\", \"newS7\")]\n",
    "# lib_comparisons_to_test += [(\"newerN2\", \"oldN2\")]\n",
    "# lib_comparisons_to_test += [(\"sPM57\", \"sPM58\")]\n",
    "\n",
    "\n",
    "gene_lists = {}\n",
    "\n",
    "for libs in lib_comparisons_to_test:\n",
    "    first_lib, second_lib = libs\n",
    "    # Run the tests:\n",
    "    with warnings.catch_warnings():\n",
    "        if \"chi2\" in tests_to_run:\n",
    "            tqdm.pandas(desc=f\"Calculating Chi Squared for {first_lib} and {second_lib}\")\n",
    "            filter_df[[f\"{first_lib}_v_{second_lib}_chi2_test_result\", f\"{first_lib}_v_{second_lib}_chi2_p_value\"]] = filter_df.progress_apply(lambda row: row_chi2(row, first_lib, second_lib), axis=1, result_type=\"expand\")\n",
    "        if \"fishers\" in tests_to_run:\n",
    "            tqdm.pandas(desc=f\"Calculating Fisher's exact for {first_lib} and {second_lib}\")\n",
    "            filter_df[[f\"{first_lib}_v_{second_lib}_fishers_test_result\", f\"{first_lib}_v_{second_lib}_fishers_p_value\"]] = filter_df.progress_apply(lambda row: row_fishers_exact(row, first_lib, second_lib, hits_or_rpm='hits', alternative='less'), axis=1, result_type=\"expand\")\n",
    "        \n",
    "        # Barnard and Boschloo take so so long to run and don't provide much of a power increase...\n",
    "        if \"barnard\" in tests_to_run:\n",
    "            tqdm.pandas(desc=f\"Calculating Barnard's exact for {first_lib} and {second_lib}\")\n",
    "            filter_df[[f\"{first_lib}_v_{second_lib}_barnard_test_result\", f\"{first_lib}_v_{second_lib}_barnard_p_value\"]] = filter_df.progress_apply(lambda row: row_barnard_exact(row, first_lib, second_lib, hits_or_rpm='hits', alternative='less', sampling_points=4), axis=1, result_type=\"expand\")\n",
    "        if \"boschloo\" in tests_to_run:\n",
    "            tqdm.pandas(desc=f\"Calculating Boschloo exact for {first_lib} and {second_lib}\")\n",
    "            filter_df[[f\"{first_lib}_v_{second_lib}_boschloo_test_result\", f\"{first_lib}_v_{second_lib}_boschloo_p_value\"]] = filter_df.progress_apply(lambda row: row_boschloo_exact(row, first_lib, second_lib, hits_or_rpm='hits', alternative='less', sampling_points=4), axis=1, result_type=\"expand\")\n",
    "    # Make adjustments for multiple testing:\n",
    "    for stat_test in tests_to_run:\n",
    "        lib_cols_for_correction = []\n",
    "        for lib in libs:\n",
    "            filter_col_converter = {\"total\": f\"total_gene_{filter_with_rpm_or_hits}_{lib}\",\n",
    "                                    \"adapted\": f\"gene_{filter_with_rpm_or_hits}_{lib}_t5+\",\n",
    "                                    \"unadapted\": f\"gene_{filter_with_rpm_or_hits}_{lib}_t5-\"}\n",
    "            lib_cols_for_correction.append(filter_col_converter[filter_col_target])\n",
    "        cumulative_col_name = f\"cumulative_{filter_col_target}_{filter_with_rpm_or_hits}_{first_lib}_{second_lib}\"\n",
    "        filter_df[cumulative_col_name] = filter_df[lib_cols_for_correction[0]] + filter_df[lib_cols_for_correction[1]]\n",
    "        number_of_genes_passing_cutoff = filter_df[filter_df[cumulative_col_name] >= cumulative_min_read_cutoff].shape[0]\n",
    "        passed_cutoff_df = filter_df[filter_df[cumulative_col_name] >= cumulative_min_read_cutoff]\n",
    "        gene_lists[(libs, stat_test)] = passed_cutoff_df.index\n",
    "        adjusted_sig_cutoff = base_sig_cutoff / number_of_genes_passing_cutoff\n",
    "        p_value_cutoff_dict[(libs, stat_test)] = (adjusted_sig_cutoff, number_of_genes_passing_cutoff)\n",
    "        \n",
    "        print_and_write(f\"There were {number_of_genes_passing_cutoff} genes that passed the cutoff of having >={cumulative_min_read_cutoff} cumulative {filter_col_target} {filter_with_rpm_or_hits} between {first_lib} and {second_lib}\"\n",
    "              f\"\\n\\tA Bonferroni correction with this in mind will expect a p value of {adjusted_sig_cutoff:.3g} for a significant {stat_test} test result.\")\n",
    "        print_and_write(f\"***But I think this isn't a great way to filter, I am going to recalculate the cutoff based on the number of genes that pass the filter for ALL LIBRARIES TESTED... This does mean the pool of libs tested matters a lot more!***\")\n",
    "# Now we need to recalculate the cutoff based on the number of genes that pass the filter for ALL LIBRARIES TESTED:\n",
    "# First we'll create the tested_genes_df and remove all duplicates, then we'll calculate the adjusted cutoff:\n",
    "stat_test_sig_cutoffs = {}\n",
    "stat_tested_genes_dfs = {}\n",
    "for stat_test in tests_to_run:\n",
    "    tested_genes_df = pd.DataFrame(columns=['gene_id', 'gene_name'])\n",
    "    for (libs, list_stat_test), gene_list in gene_lists.items():\n",
    "        if stat_test != list_stat_test:\n",
    "            continue\n",
    "        lib1, lib2 = libs\n",
    "        gene_list_better = gene_list.to_frame().reset_index(drop=True)[[\"gene_id\", \"gene_name\"]]\n",
    "        gene_list_better[f'tested_for_{stat_test}_{lib1}-v-{lib2}'] = True\n",
    "        tested_genes_df = pd.merge(tested_genes_df, gene_list_better, how='outer', on=['gene_id', 'gene_name'])\n",
    "        tested_genes_df = tested_genes_df.fillna(False)\n",
    "        tested_genes_df = tested_genes_df.drop_duplicates()\n",
    "    tested_genes_df[f'tested_{stat_test}_for_all'] = tested_genes_df.all(axis='columns')\n",
    "    total_genes_tested_count = tested_genes_df.shape[0]\n",
    "    universally_tested_genes_count = tested_genes_df[tested_genes_df[f'tested_{stat_test}_for_all']].shape[0]\n",
    "    adjusted_sig_cutoff = base_sig_cutoff / universally_tested_genes_count\n",
    "    stat_test_sig_cutoffs[stat_test] = adjusted_sig_cutoff\n",
    "    stat_tested_genes_dfs[stat_test] = tested_genes_df\n",
    "    print_and_write(f\"Of the {total_genes_tested_count} genes tested for {stat_test}, {universally_tested_genes_count} were tested among all libraries.\")\n",
    "    # Add a column to the filter_df, that stores a Bool depending on if the gene was in the tested_genes_df[f'tested_{stat_test}_for_all'] column:\n",
    "    temp_filter_df = filter_df.copy().reset_index()\n",
    "    temp_filter_df = pd.merge(temp_filter_df, tested_genes_df[['gene_id', 'gene_name', f'tested_{stat_test}_for_all']], how='left', on=['gene_id', 'gene_name'])\n",
    "    temp_filter_df[f'tested_{stat_test}_for_all'] = temp_filter_df[f'tested_{stat_test}_for_all'].fillna(False)\n",
    "    filter_df = temp_filter_df.set_index(['chr_id', 'gene_id', 'gene_name'])\n",
    "\n",
    "\n",
    "for stat_test in tests_to_run:\n",
    "    for (first_lib, second_lib) in lib_comparisons_to_test:\n",
    "        try:\n",
    "            filter_df.sort_values(f\"{first_lib}_v_{second_lib}_{stat_test}_p_value\")\n",
    "            filter_df[f\"neg_log10_{first_lib}_v_{second_lib}_{stat_test}_p_value\"] = -np.log10(filter_df[f\"{first_lib}_v_{second_lib}_{stat_test}_p_value\"])\n",
    "            filter_df[f\"{first_lib}_v_{second_lib}_{stat_test}_significant\"] = filter_df[f\"{first_lib}_v_{second_lib}_{stat_test}_p_value\"] <= stat_test_sig_cutoffs[stat_test]\n",
    "            # Some nice printing!\n",
    "            #####################\n",
    "            print_and_write(f\"For {first_lib} v {second_lib} {stat_test} test:\")\n",
    "            print_and_write(f\"    {filter_df[filter_df[f'{first_lib}_v_{second_lib}_{stat_test}_significant']].shape[0]} genes were significant at a p value of {stat_test_sig_cutoffs[stat_test]:.3g} (Bonferroni corrected for all genes tested)\")\n",
    "            sig_genes = filter_df[filter_df[f'{first_lib}_v_{second_lib}_{stat_test}_significant']][[f'{first_lib}_v_{second_lib}_{stat_test}_p_value']].reset_index()[['gene_id', 'gene_name', f'{first_lib}_v_{second_lib}_{stat_test}_p_value']].sort_values(f'{first_lib}_v_{second_lib}_{stat_test}_p_value')\n",
    "            sig_genes_list = sig_genes.values.tolist()\n",
    "            sig_genes_string = '\\n'.join([f\"        {gene_id}    {gene_name:<10}    p-val: {p_value:.3g}\" for gene_id, gene_name, p_value in sig_genes_list])\n",
    "            print_and_write(f\"    Genes:\\n{sig_genes_string}\")\n",
    "            #####################\n",
    "            \n",
    "        except KeyError:\n",
    "            print(f\"Couldn't find columns corresponding to '{stat_test}'!! Be sure spelling is correct!\")\n",
    "    print(f\"Total Count of genes tested: {filter_df.shape[0]}, with {filter_df[filter_df[f'tested_{stat_test}_for_all']].shape[0]} genes tested among all libraries.\")\n",
    "# Let's just roll back to a simple save, writing the entire df to a csv\n",
    "df_output_path = stat_test_record_file.parent / f\"{npCommon.get_dt()}_statTests_{'-'.join(tests_to_run)}_largeDF.csv\"\n",
    "print_and_write(f\"Saving the entire df to {df_output_path}...\")\n",
    "filter_df.to_csv(df_output_path)\n",
    "print_and_write('done.')"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Look at the top \"most significant\" hits"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n",
    "how_many_hits_to_save = None\n",
    "load_and_merge_simple = True\n",
    "for stat_test in tests_to_run:\n",
    "    for first_lib, second_lib in lib_comparisons_to_test:\n",
    "        temp_df = filter_df[[f\"gene_rpm_{first_lib}_t5-\",\n",
    "                            f\"gene_rpm_{first_lib}_t5+\",\n",
    "                            f\"gene_rpm_{second_lib}_t5-\",\n",
    "                            f\"gene_rpm_{second_lib}_t5+\",\n",
    "                            f\"{first_lib}_v_{second_lib}_{stat_test}_test_result\",\n",
    "                            f\"{first_lib}_v_{second_lib}_{stat_test}_p_value\",\n",
    "                            f\"{first_lib}_v_{second_lib}_{stat_test}_significant\",\n",
    "                            ]].sort_values(f\"{first_lib}_v_{second_lib}_{stat_test}_p_value\",\n",
    "                                           ascending=True)\n",
    "        save_path = f\"./output_files/{npCommon.get_dt()}_{first_lib}_v_{second_lib}.{stat_test}\"\n",
    "        if isinstance(how_many_hits_to_save, int):\n",
    "            temp_df = temp_df.head(how_many_hits_to_save)\n",
    "            save_path += f\".top{how_many_hits_to_save}\"\n",
    "        save_path += \".csv\"\n",
    "        print(temp_df)\n",
    "        temp_df.to_csv(save_path)\n",
    "        print(f\"Saved top {how_many_hits_to_save} hits to {save_path}\")\n",
    "if load_and_merge_simple:\n",
    "    df_dict_fishers = {}\n",
    "    for first_lib, second_lib in lib_comparisons_to_test:\n",
    "        load_path = f\"./output_files/{npCommon.get_dt()}_{first_lib}_v_{second_lib}.fishers\"\n",
    "        if isinstance(how_many_hits_to_save, int):\n",
    "            load_path += f\".top{how_many_hits_to_save}\"\n",
    "        load_path += \".csv\"\n",
    "        df_dict_fishers[(first_lib, second_lib)] = pd.read_csv(load_path)[['gene_name', 'gene_id', f\"{first_lib}_v_{second_lib}_fishers_p_value\", f\"{first_lib}_v_{second_lib}_fishers_significant\"]]\n",
    "    # Next I want to merge all of these dataframes in df_dict_fishers into one based on the gene_id and gene_name columns:\n",
    "    df_dict_fishers_merged = pd.DataFrame(columns=['gene_name', 'gene_id'])\n",
    "    for first_lib, second_lib in lib_comparisons_to_test:\n",
    "        df_dict_fishers_merged = df_dict_fishers_merged.merge(df_dict_fishers[(first_lib, second_lib)], how='outer', on=['gene_name', 'gene_id'])\n",
    "    print(df_dict_fishers_merged.head())\n",
    "    df_dict_fishers_merged.to_csv(f\"./output_files/{npCommon.get_dt()}_merged_fishers.csv\")\n",
    "df_dict_fishers_merged"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Plot fraction adapted fold change and Chi2/Fishers p-values"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-06-29T17:25:19.672449968Z",
     "start_time": "2023-06-29T17:25:18.055101100Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "       gene_name                prev_identified_cols  prev_identified_count\n0          tos-1  ramani2009,muir2018,kim_modena2022                      3\n1          odc-1  ramani2009,muir2018,kim_modena2022                      3\n2      Y73B3A.18             muir2018,kim_modena2022                      2\n3          smd-1  ramani2009,muir2018,kim_modena2022                      3\n4        C53H9.2  ramani2009,muir2018,kim_modena2022                      3\n...          ...                                 ...                    ...\n16543    ZK896.3                                                          0\n16544    ZK896.5                                                          0\n16545   clec-186                                                          0\n16546     nstp-5                                                          0\n16547      abf-2                                                          0\n\n[16548 rows x 3 columns]",
      "text/html": "<div>\n<style scoped>\n    .dataframe tbody tr th:only-of-type {\n        vertical-align: middle;\n    }\n\n    .dataframe tbody tr th {\n        vertical-align: top;\n    }\n\n    .dataframe thead th {\n        text-align: right;\n    }\n</style>\n<table border=\"1\" class=\"dataframe\">\n  <thead>\n    <tr style=\"text-align: right;\">\n      <th></th>\n      <th>gene_name</th>\n      <th>prev_identified_cols</th>\n      <th>prev_identified_count</th>\n    </tr>\n  </thead>\n  <tbody>\n    <tr>\n      <th>0</th>\n      <td>tos-1</td>\n      <td>ramani2009,muir2018,kim_modena2022</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>1</th>\n      <td>odc-1</td>\n      <td>ramani2009,muir2018,kim_modena2022</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>2</th>\n      <td>Y73B3A.18</td>\n      <td>muir2018,kim_modena2022</td>\n      <td>2</td>\n    </tr>\n    <tr>\n      <th>3</th>\n      <td>smd-1</td>\n      <td>ramani2009,muir2018,kim_modena2022</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>4</th>\n      <td>C53H9.2</td>\n      <td>ramani2009,muir2018,kim_modena2022</td>\n      <td>3</td>\n    </tr>\n    <tr>\n      <th>...</th>\n      <td>...</td>\n      <td>...</td>\n      <td>...</td>\n    </tr>\n    <tr>\n      <th>16543</th>\n      <td>ZK896.3</td>\n      <td></td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>16544</th>\n      <td>ZK896.5</td>\n      <td></td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>16545</th>\n      <td>clec-186</td>\n      <td></td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>16546</th>\n      <td>nstp-5</td>\n      <td></td>\n      <td>0</td>\n    </tr>\n    <tr>\n      <th>16547</th>\n      <td>abf-2</td>\n      <td></td>\n      <td>0</td>\n    </tr>\n  </tbody>\n</table>\n<p>16548 rows × 3 columns</p>\n</div>"
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "execution_count": 83
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "from plotly.subplots import make_subplots\n",
    "def plot_multiple_test_scatters(multi_plot_df, lib_sets_to_plot, stat_test_to_plot, save_path_for_fig, round_inf_to=8, filter_based_on_all_not_lib_alone=True, shapes_col=None):\n",
    "    for first_lib, second_lib in lib_sets_to_plot:\n",
    "        multi_plot_df[f\"fracAdapted_FC_{first_lib}/{second_lib}\"] = np.log2(\n",
    "            multi_plot_df[f\"fraction_adapted_{first_lib}\"] / multi_plot_df[f\"fraction_adapted_{second_lib}\"])\n",
    "    multi_plot_df = multi_plot_df.reset_index()\n",
    "    multi_plot_df = multi_plot_df[multi_plot_df.chr_id != 'MtDNA']\n",
    "    multi_plot_df = multi_plot_df.replace({-np.inf: -round_inf_to, np.inf: round_inf_to})\n",
    "    \n",
    "    fig = make_subplots(rows=1, cols=len(lib_sets_to_plot),\n",
    "                        shared_yaxes=True,\n",
    "                        shared_xaxes=True,\n",
    "                        vertical_spacing=0.02,\n",
    "                        horizontal_spacing=0.02,\n",
    "                        # subplot_titles=[f\"{libs[0]} vs {libs[1]}\" for libs in lib_comparisons_to_plot],\n",
    "                        )\n",
    "    for i, (first_lib, second_lib) in enumerate(lib_sets_to_plot):\n",
    "        print(first_lib, second_lib)\n",
    "        x_col = f\"fracAdapted_FC_{first_lib}/{second_lib}\"\n",
    "        y_col = f\"neg_log10_{first_lib}_v_{second_lib}_{stat_test_to_plot}_p_value\"\n",
    "        sig_col = f\"{first_lib}_v_{second_lib}_{stat_test_to_plot}_significant\"\n",
    "        \n",
    "        if shapes_col is None:\n",
    "            marker_symbol_col = None\n",
    "        elif multi_plot_df[shapes_col].dtype == bool:\n",
    "            multi_plot_df[f'{shapes_col}_{first_lib}/{second_lib}'] = multi_plot_df[shapes_col] & multi_plot_df[sig_col]\n",
    "            marker_symbol_col = multi_plot_df[f'{shapes_col}_{first_lib}/{second_lib}'].replace({False: 'circle', True: 'cross'})\n",
    "        elif multi_plot_df[shapes_col].dtype == int:\n",
    "            new_shapes_col = f'{shapes_col}_{first_lib}/{second_lib}'\n",
    "            multi_plot_df[new_shapes_col] = multi_plot_df[shapes_col]\n",
    "            multi_plot_df.loc[~multi_plot_df[sig_col], new_shapes_col] = 0\n",
    "            marker_symbol_col = multi_plot_df[new_shapes_col].replace({0: 'circle', 1: 'cross', 2: 'triangle-up', 3: 'square'})\n",
    "        else:\n",
    "            raise TypeError(f\"shapes_col must be a bool or int dtype, not {multi_plot_df[shapes_col].dtype}:\\n\\n{multi_plot_df[shapes_col].head()}\")\n",
    "        \n",
    "        \n",
    "        fig.add_trace(go.Scatter(\n",
    "            x=multi_plot_df[x_col],\n",
    "            y=multi_plot_df[y_col],\n",
    "            marker_color=multi_plot_df[sig_col].replace({False: \"black\", True: \"red\"}),\n",
    "            marker_symbol=marker_symbol_col,\n",
    "            hovertext=multi_plot_df[\"gene_name\"],\n",
    "            # start GitHub copilot help. It made this HELLA easy!!:\n",
    "            customdata=multi_plot_df[[f\"total_gene_rpm_{first_lib}\", f\"total_gene_rpm_{second_lib}\", \"prev_id_list\"]],\n",
    "            hovertemplate=\"<b>%{hovertext}</b><br>\" +\n",
    "                          \"FC Fraction Adapted: %{x:.2f}<br>\" +\n",
    "                          \"-log<sub>10</sub> P-Value: %{y:.2e}<br>\" +\n",
    "                          f\"total gene RPM {first_lib}: \" + \"%{customdata[0]:.2f}<br>\" +\n",
    "                          f\"total gene RPM {second_lib}: \" + \"%{customdata[1]:.2f}<br>\" +\n",
    "                          \"prev_id_list: %{customdata[2]}<br>\" +\n",
    "                          \"<extra></extra>\",\n",
    "            # end GitHub copilot help\n",
    "            mode=\"markers\",\n",
    "            name=f\"{second_lib}\"),\n",
    "            row=1, col=i + 1)\n",
    "        if filter_based_on_all_not_lib_alone:\n",
    "            cutoff = stat_test_sig_cutoffs[stat_test]\n",
    "            num_passed_genes = multi_plot_df.tested_fishers_for_all.sum()\n",
    "        else:\n",
    "            cutoff, num_passed_genes = p_value_cutoff_dict[((first_lib, second_lib), stat_test_to_plot)]\n",
    "        fig.add_hline(y=-np.log10(cutoff),\n",
    "                      row=1, col=i + 1)\n",
    "        fig.add_annotation(text=f\"Bonferroni<br>p-value = {cutoff:0.3e}\",\n",
    "                           yref=\"y\",\n",
    "                           y=-np.log10(cutoff),\n",
    "                           xref=\"x domain\",\n",
    "                           x=0,\n",
    "                           yanchor=\"bottom\",\n",
    "                           showarrow=False,\n",
    "                           align=\"left\",\n",
    "                           row=1, col=i + 1)\n",
    "        \n",
    "        # Bottom of plot text:\n",
    "        bottom_text = f\"<b>genes w/ > 100 {filter_col_target} {filter_with_rpm_or_hits} = {num_passed_genes}\"\n",
    "        fig.add_annotation(text=bottom_text,\n",
    "                           yref=\"y domain\",\n",
    "                           y=0,\n",
    "                           xref=\"x domain\",\n",
    "                           x=0.5,\n",
    "                           yanchor=\"bottom\",\n",
    "                           showarrow=False,\n",
    "                           align=\"center\",\n",
    "                           row=1, col=i + 1)\n",
    "        # Cutoffs to isolate the +INF genes:\n",
    "        fig.add_vrect(x0=round_inf_to - 1.0,\n",
    "                      x1=round_inf_to - 0.5,\n",
    "                      line_width=0,\n",
    "                      fillcolor='white',\n",
    "                      )\n",
    "        fig.add_vrect(x0=round_inf_to + 1.0,\n",
    "                      x1=round_inf_to + 0.5,\n",
    "                      line_width=0,\n",
    "                      fillcolor='white',\n",
    "                      )\n",
    "        # X Axis Title:\n",
    "        fig.update_xaxes(title_text=f\"Fold Change Fraction Adapted<br>N2 vs. smg-{second_lib[-1]} ({second_lib[:-2]})\",\n",
    "                         row=1, col=i + 1)\n",
    "    fig.update_layout(\n",
    "        font=dict(\n",
    "            family=\"Rockwell\",  # \"Rockwell\" \"Courier New, monospace\"\n",
    "            size=16,  # Set the font size here\n",
    "            color=\"black\"\n",
    "        ),\n",
    "        hoverlabel=dict(\n",
    "            bgcolor=\"white\",\n",
    "            font_size=20,\n",
    "            font_family=\"Rockwell\",  # \"Rockwell\" \"Courier New, monospace\"\n",
    "        ),\n",
    "        showlegend=False,\n",
    "    )\n",
    "    # Title:\n",
    "    title_text = (f\"<b>FC Frac. Adapted vs. {stat_test_to_plot.title()} Test P-Values</b>\"\n",
    "                 f\"<br>Cumulative {filter_col_target} {filter_with_rpm_or_hits} for Cutoff = {cumulative_min_read_cutoff}\")\n",
    "    if shapes_col is not None:\n",
    "        title_text += f\"<br>Cross = {shapes_col}\"\n",
    "    fig.update_layout(template=\"plotly_white\",\n",
    "                      title=title_text)\n",
    "    fig.update_xaxes(tickvals=[f\"-{round_inf_to}\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               f\"{round_inf_to}\"],\n",
    "                     ticktext=[\"-INF\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               \"INF\"],\n",
    "                     range=[-2, round_inf_to+0.5])\n",
    "    fig.update_yaxes(title_text=f\"-log<sub>10</sub> {stat_test_to_plot.title()} Test P-Value\",\n",
    "                     row=1, col=1)\n",
    "\n",
    "    fig.write_html(save_path_for_fig + \".html\")\n",
    "    fig.update_layout(\n",
    "        autosize=False,\n",
    "        width=600*len(lib_sets_to_plot),\n",
    "        height=600, )\n",
    "    fig.write_image(save_path_for_fig + \".svg\")\n",
    "    fig.write_image(save_path_for_fig + \".png\")\n",
    "    fig.show()\n",
    "    return multi_plot_df\n",
    "\n",
    "stat_test = \"fishers\"  # \"fishers\" or \"chi2\"\n",
    "lib_comparisons_to_plot = [\n",
    "    # (\"newN2\", \"newS5\"),\n",
    "    # (\"newN2\", \"newS6\"),\n",
    "    # (\"newN2\", \"newS7\"),\n",
    "    (\"oldN2\", \"oldS6\"),\n",
    "    # (\"sPM57\", \"sPM58\"),\n",
    "    (\"newerN2\", \"newerS6\"),\n",
    "    (\"newerN2\", \"newerS5\"),\n",
    "    (\"newN2\", \"newS7\"),\n",
    "    # (\"newerN2\", \"oldN2\"),\n",
    "]\n",
    "flat_plot_libs = [item for sublist in lib_comparisons_to_plot for item in sublist]\n",
    "flat_plot_libs = sorted(list(set(flat_plot_libs)))\n",
    "save_dir = Path(f\"{output_dir}\")\n",
    "save_name =  f\"{npCommon.get_dt()}_{'-'.join(flat_plot_libs)}_{stat_test}Pvalues-v-FCfractionAdapted.scatters\"\n",
    "plot_df = filter_df.copy(deep=True)\n",
    "# plot_df = plot_df[plot_df.tested_fishers_for_all]\n",
    "\n",
    "### Adding on 6/27/2023 to add shapes depending on previously identified NMD targets:\n",
    "marker_shape_col = \"prev_id_count\" # ramani2009_hit, muir2018_hit, kim_modena2022_smg6_hit, kim_modena2022_smg1_hit, or mitrovich_hit\n",
    "prev_targets_df = pd.read_csv(\"/data16/marcus/working/230602_comparingNMDTargets/230627_totalMerge.csv\")\n",
    "prev_targets_df['kim_modena2022_hit'] = (prev_targets_df['kim_modena2022_smg1_hit'] | prev_targets_df['kim_modena2022_smg6_hit'])\n",
    "prev_targets_df['prev_id_count'] = prev_targets_df[['muir2018_hit', 'ramani2009_hit', 'kim_modena2022_hit']].sum(axis=1)\n",
    "prev_targets_df['prev_id_list'] = prev_targets_df[['muir2018_hit', 'ramani2009_hit', 'kim_modena2022_hit']].apply(lambda row: \", \".join([col.rstrip(\"_hit\") for col in row.index if row[col]]), axis=1)\n",
    "plot_df = plot_df.reset_index()\n",
    "# Drop columns that show up in prev_targets AND plot_df\n",
    "plot_df = plot_df.merge(prev_targets_df[['gene_name', 'gene_id', marker_shape_col, 'prev_id_list']], on=['gene_name', 'gene_id'], how=\"left\").set_index(['chr_id', 'gene_id', 'gene_name'])\n",
    "if marker_shape_col == \"prev_id_count\":\n",
    "    plot_df[marker_shape_col] = plot_df[marker_shape_col].fillna(0)\n",
    "    plot_df[marker_shape_col] = plot_df[marker_shape_col].astype(int)\n",
    "else:\n",
    "    plot_df[marker_shape_col] = plot_df[marker_shape_col].fillna(False)\n",
    "save_name += f\".withShapesFrom_{marker_shape_col}\"\n",
    "###\n",
    "\n",
    "save_path = save_dir / save_name\n",
    "plot_df = plot_multiple_test_scatters(plot_df, lib_comparisons_to_plot, stat_test, str(save_path), shapes_col=marker_shape_col, filter_based_on_all_not_lib_alone=True)\n",
    "\n",
    "plot_df.query(\"gene_name == 'F19B2.5'\").T"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Looking at reproducibility\n",
    "## April 3, 2023"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2023-06-29T17:31:30.784392484Z",
     "start_time": "2023-06-29T17:31:27.494729902Z"
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/lib/python3.8/dist-packages/pandas/core/arraylike.py:402: RuntimeWarning:\n",
      "\n",
      "divide by zero encountered in log2\n",
      "\n",
      "/usr/local/lib/python3.8/dist-packages/pandas/core/arraylike.py:402: RuntimeWarning:\n",
      "\n",
      "divide by zero encountered in log2\n",
      "\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "newerN2 newerS6\n",
      "newerN2 newerS5\n"
     ]
    }
   ],
   "execution_count": 84
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n",
    "lib_comparisons_to_compare = [  # Try to just do two for now!\n",
    "    # (\"newN2\", \"newS5\"),\n",
    "    # (\"newN2\", \"newS6\"),\n",
    "    # (\"newN2\", \"newS7\"),\n",
    "    # (\"oldN2\", \"oldS6\"),\n",
    "    (\"newerN2\", \"newerS6\"),\n",
    "    (\"oldN2\", \"oldS6\"),\n",
    "    # (\"newerN2\", \"newerS5\"),\n",
    "    # (\"newN2\", \"newS7\"),\n",
    "]\n",
    "stat_test = 'fishers'  # 'fishers' or 'chi2'\n",
    "\n",
    "(x_lib1, x_lib2), (y_lib1, y_lib2) = x_libs, y_libs = lib_comparisons_to_compare\n",
    "\n",
    "lib_comparisons_to_filter = [f\"{libs[0]}_v_{libs[1]}\" for libs in lib_comparisons_to_compare]\n",
    "filter_cols = [col for col in plot_df.columns if any([filter_substring in col for filter_substring in lib_comparisons_to_filter]) or col in ['chr_id', 'gene_id', 'gene_name']]\n",
    "filter_cols += [col for col in plot_df.columns if any([lib_name in col for sublist in lib_comparisons_to_compare for lib_name in sublist])]\n",
    "filter_cols = set(filter_cols)\n",
    "print(filter_cols)\n",
    "filter_cols = [col for col in filter_cols if not \"chi2\" in col]\n",
    "new_df = plot_df[filter_cols]\n",
    "\n",
    "tested_genes_list = stat_tested_genes_dfs[stat_test][stat_tested_genes_dfs[stat_test][f'tested_{stat_test}_for_all']].gene_id.to_list()\n",
    "new_df = new_df.query(f\"gene_id in @tested_genes_list\")\n",
    "sea.set_style('white')\n",
    "\n",
    "# npCommon.boolDF_to_upsetPlot(new_df)\n",
    "# new_new_df = new_df.select_dtypes(include='bool')\n",
    "# new_new_df[new_new_df[new_new_df.columns[0]] + new_new_df[new_new_df.columns[1]]].merge(new_df['gene_name'], right_index=True, left_index=True)\n",
    "print(new_df.columns)\n",
    "x_col = f'neg_log10_{x_lib1}_v_{x_lib2}_fishers_p_value'\n",
    "y_col = f'neg_log10_{y_lib1}_v_{y_lib2}_fishers_p_value'\n",
    "# new_df = new_df[~new_df[x_col].isna()]\n",
    "# new_df = new_df[~new_df[y_col].isna()]\n",
    "new_df['average_rpm'] = (new_df[f'total_gene_rpm_{x_lib1}'] +\n",
    "                         new_df[f'total_gene_rpm_{x_lib2}'] +\n",
    "                         new_df[f'total_gene_rpm_{y_lib1}'] +\n",
    "                         new_df[f'total_gene_rpm_{y_lib2}']) / 4 + 1\n",
    "rpm_dict = {f'total_gene_rpm_{lib}': f'Gene RPM ({lib})' for lib in [x_lib1, x_lib2, y_lib1, y_lib2]}\n",
    "labels_dict = {\n",
    "                     x_col: f\"-log<sub>10</sub> {stat_test.title()} Test P-Value ({x_lib1} vs {x_lib2})\",\n",
    "                     y_col: f\"-log<sub>10</sub> {stat_test.title()} Test P-Value ({y_lib1} vs {y_lib2})\",\n",
    "                     \"chr_id\": \"Chromosome\",\n",
    "                 }\n",
    "labels_dict.update(rpm_dict)\n",
    "# Just drop the freaking weird pseudogenes: F23A7.4 and .8:\n",
    "new_df = new_df[~new_df.gene_name.isin(['F23A7.4', 'F23A7.8'])]\n",
    "print(new_df.head())\n",
    "fig = px.scatter(new_df[new_df.average_rpm > 150],\n",
    "                 x=x_col,\n",
    "                 y=y_col,\n",
    "                 hover_name='gene_name',\n",
    "                 hover_data=['chr_id'] + list(rpm_dict),\n",
    "                 template='none',\n",
    "                 # size='average_rpm',\n",
    "                 height=1000,\n",
    "                 width=1000,\n",
    "                 labels=labels_dict,\n",
    "                 )\n",
    "fig.update_traces(marker=dict(color='black',\n",
    "                              size=10))\n",
    "x_cutoff, _ = p_value_cutoff_dict[(x_libs, stat_test)]\n",
    "y_cutoff, _ = p_value_cutoff_dict[(y_libs, stat_test)]\n",
    "fig.add_shape(y0=-np.log10(y_cutoff), y1=-np.log10(y_cutoff), yref='y',\n",
    "              x0=0, x1=1, xref='paper',\n",
    "              type='line', line_color='salmon', line_width=3, line_dash='dot',\n",
    "              opacity=1,\n",
    "              )\n",
    "fig.add_shape(x0=-np.log10(y_cutoff), x1=-np.log10(y_cutoff), xref='x',\n",
    "              y0=0, y1=1, yref='paper',\n",
    "              type='line', line_color='salmon', line_width=3, line_dash='dot',\n",
    "              opacity=1,\n",
    "              )\n",
    "fig.update_yaxes(title_text=f\"<b>-log<sub>10</sub> {stat_test.title()} Test P-Value<br>\"\n",
    "                            f\"{y_lib1} vs {y_lib2}\")\n",
    "fig.update_xaxes(title_text=f\"<b>-log<sub>10</sub> {stat_test.title()} Test P-Value<br>\"\n",
    "                            f\"{x_lib1} vs {x_lib2}\")\n",
    "fig.update_yaxes(rangemode='nonnegative')\n",
    "fig.update_xaxes(rangemode='nonnegative')\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Courier New, monospace\",\n",
    "        size=18,  # Set the font size here\n",
    "        color=\"black\"\n",
    "    ),\n",
    "     hoverlabel=dict(\n",
    "        bgcolor=\"white\",\n",
    "        font_size=20,\n",
    "        font_family=\"Rockwell\"\n",
    "    )\n",
    ")\n",
    "fig.update_layout(yaxis = dict( tickfont = dict(size=20)),\n",
    "                  xaxis = dict( tickfont = dict(size=20)))\n",
    "save_dir = \"/home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python\"\n",
    "save_path = save_dir + f\"/{npCommon.get_dt()}_-log10{stat_test.title()}_{y_lib1.rstrip('N2')}_v_{x_lib1.rstrip('N2')}\"\n",
    "fig.write_html(save_path + \".html\")\n",
    "fig.write_image(save_path + \".svg\")\n",
    "fig.write_image(save_path + \".png\")\n",
    "fig.show()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Plot the relationship between fraction adapted between smg-5 and smg-6\n",
    "This will go with the similar plots of P-values"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "\n",
    "lib_comparisons_to_compare = [  # Try to just do two for now!\n",
    "    # (\"newN2\", \"newS5\"),\n",
    "    # (\"newN2\", \"newS6\"),\n",
    "    # (\"newN2\", \"newS7\"),\n",
    "    # (\"oldN2\", \"oldS6\"),\n",
    "    (\"newerN2\", \"newerS6\"),\n",
    "    (\"newerN2\", \"newerS5\"),\n",
    "]\n",
    "stat_test = 'fishers'  # 'fishers' or 'chi2'\n",
    "\n",
    "(x_lib1, x_lib2), (y_lib1, y_lib2) = x_libs, y_libs = lib_comparisons_to_compare\n",
    "\n",
    "lib_comparisons_to_filter = [f\"{libs[0]}_v_{libs[1]}\" for libs in lib_comparisons_to_compare]\n",
    "filter_cols = [col for col in plot_df.columns if any([filter_substring in col for filter_substring in lib_comparisons_to_filter]) or col in ['chr_id', 'gene_id', 'gene_name']]\n",
    "filter_cols += [col for col in plot_df.columns if any([lib_name in col for sublist in lib_comparisons_to_compare for lib_name in sublist])]\n",
    "filter_cols = set(filter_cols)\n",
    "print(filter_cols)\n",
    "filter_cols = [col for col in filter_cols if not \"chi2\" in col]\n",
    "new_df = plot_df[filter_cols]\n",
    "\n",
    "tested_genes_list = stat_tested_genes_dfs[stat_test][stat_tested_genes_dfs[stat_test][f'tested_{stat_test}_for_all']].gene_id.to_list()\n",
    "new_df = new_df.query(f\"gene_id in @tested_genes_list\")\n",
    "sea.set_style('white')\n",
    "\n",
    "# npCommon.boolDF_to_upsetPlot(new_df)\n",
    "# new_new_df = new_df.select_dtypes(include='bool')\n",
    "# new_new_df[new_new_df[new_new_df.columns[0]] + new_new_df[new_new_df.columns[1]]].merge(new_df['gene_name'], right_index=True, left_index=True)\n",
    "print(new_df.columns)\n",
    "# x_col = f'neg_log10_{x_lib1}_v_{x_lib2}_fishers_p_value'\n",
    "x_col = f\"fracAdapted_FC_{x_lib1}/{x_lib2}\"\n",
    "# y_col = f'neg_log10_{y_lib1}_v_{y_lib2}_fishers_p_value'\n",
    "y_col = f\"fracAdapted_FC_{y_lib1}/{y_lib2}\"\n",
    "# new_df = new_df[~new_df[x_col].isna()]\n",
    "# new_df = new_df[~new_df[y_col].isna()]\n",
    "new_df['average_rpm'] = (new_df[f'total_gene_rpm_{x_lib1}'] +\n",
    "                         new_df[f'total_gene_rpm_{x_lib2}'] +\n",
    "                         new_df[f'total_gene_rpm_{y_lib1}'] +\n",
    "                         new_df[f'total_gene_rpm_{y_lib2}']) / 4 + 1\n",
    "rpm_dict = {f'total_gene_rpm_{lib}': f'Gene RPM ({lib})' for lib in [x_lib1, x_lib2, y_lib1, y_lib2]}\n",
    "labels_dict = {\n",
    "                     x_col: f\"Fraction Adapted Fold Change ({x_lib1} vs {x_lib2})\",\n",
    "                     y_col: f\"Fraction Adapted Fold Change ({y_lib1} vs {y_lib2})\",\n",
    "                     \"chr_id\": \"Chromosome\",\n",
    "                 }\n",
    "labels_dict.update(rpm_dict)\n",
    "# Just drop the freaking weird pseudogenes: F23A7.4 and .8:\n",
    "new_df = new_df[~new_df.gene_name.isin(['F23A7.4', 'F23A7.8'])]\n",
    "print(new_df.head())\n",
    "fig = px.scatter(new_df[new_df.average_rpm > 150],\n",
    "                 x=x_col,\n",
    "                 y=y_col,\n",
    "                 hover_name='gene_name',\n",
    "                 hover_data=['chr_id'] + list(rpm_dict),\n",
    "                 template='none',\n",
    "                 # size='average_rpm',\n",
    "                 height=1000,\n",
    "                 width=1000,\n",
    "                 labels=labels_dict,\n",
    "                 )\n",
    "# I want to color the points by if they are significant in the fishers test:\n",
    "\n",
    "\n",
    "fig.update_traces(marker=dict(color='black',\n",
    "                              size=10))\n",
    "fig.update_yaxes(rangemode='nonnegative')\n",
    "fig.update_xaxes(rangemode='nonnegative')\n",
    "fig.update_layout(\n",
    "    font=dict(\n",
    "        family=\"Courier New, monospace\",\n",
    "        size=18,  # Set the font size here\n",
    "        color=\"black\"\n",
    "    ),\n",
    "     hoverlabel=dict(\n",
    "        bgcolor=\"white\",\n",
    "        font_size=20,\n",
    "        font_family=\"Rockwell\"\n",
    "    )\n",
    ")\n",
    "round_inf_to = 8\n",
    "fig.update_xaxes(tickvals=[f\"-{round_inf_to}\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               f\"{round_inf_to}\"],\n",
    "                  ticktext=[\"-INF\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               \"INF\"],\n",
    "                  range=[-2, round_inf_to+0.5])\n",
    "fig.update_yaxes(tickvals=[f\"-{round_inf_to}\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               f\"{round_inf_to}\"],\n",
    "                  ticktext=[\"-INF\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               \"INF\"],\n",
    "                  range=[-2, round_inf_to+0.5])\n",
    "\n",
    "fig.update_layout(yaxis = dict( tickfont = dict(size=20)),\n",
    "                  xaxis = dict( tickfont = dict(size=20)))\n",
    "save_dir = \"/home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python\"\n",
    "save_path = save_dir + f\"/{npCommon.get_dt()}_fracAdaptedFC_{y_lib1.rstrip('N2')}_v_{x_lib1.rstrip('N2')}\"\n",
    "fig.write_html(save_path + \".html\")\n",
    "fig.write_image(save_path + \".svg\")\n",
    "fig.write_image(save_path + \".png\")\n",
    "fig.show()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Nested Bar Plots of Fraction Adapted FC for smg-5 and smg-6"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "small_df = new_df[(new_df.newerN2_v_newerS5_fishers_significant == True) | (new_df.newerN2_v_newerS6_fishers_significant == True)]\n",
    "\n",
    "\n",
    "# So we want to make a nested bar plot, with the two colors being the fracAdapted_FC_newerN2/newerS6 and fracAdapted_FC_newerN2/newerS5.\n",
    "# The X-axis will be all of the genes in small_df\n",
    "# The Y-axis will be the fraction adapted FC\n",
    "# We want each bar side by side for each gene, not staked:\n",
    "\n",
    "fig = px.bar(small_df,\n",
    "             x=\"gene_name\",\n",
    "             y=[\"fracAdapted_FC_newerN2/newerS6\", \"fracAdapted_FC_newerN2/newerS5\", \"fracAdapted_FC_oldN2/oldS6\"],\n",
    "             # color_discrete_sequence=[\"red\", \"blue\"],\n",
    "             template=\"none\",\n",
    "             height=500,\n",
    "             width=1000,\n",
    "             barmode=\"group\",\n",
    "             )\n",
    "\n",
    "fig.update_yaxes(tickvals=[f\"-{round_inf_to}\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               f\"{round_inf_to}\"],\n",
    "                  ticktext=[\"-INF\",\n",
    "                               \"-6\", \"-4\", \"-2\", \"0\", \"2\", \"4\", \"6\",\n",
    "                               \"INF\"],\n",
    "                  range=[-2, round_inf_to+0.5])\n",
    "# fig.update_layout(yaxis = dict( tickfont = dict(size=20)),\n",
    "#                     xaxis = dict( tickfont = dict(size=20)))\n",
    "# change y axis name:\n",
    "fig.update_yaxes(title_text=\"Fraction Adapted Fold Change<br>newerN2 vs. newerS6/newerS5\")\n",
    "\n",
    "save_dir = \"/home/marcus/Insync/mviscard@ucsc.edu/Google Drive/insync_folder/NMD_cleavage_and_deadenylation_paper/raw_figures_from_python\"\n",
    "save_path = save_dir + f\"/{npCommon.get_dt()}_fracAdaptedFC_nestedBarPlot\"\n",
    "fig.write_html(save_path + \".html\")\n",
    "fig.write_image(save_path + \".svg\")\n",
    "fig.write_image(save_path + \".png\")\n",
    "fig.show(renderer=\"browser\")"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "# Assessing p-value rank \"shape\""
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "# lib_comparisons_to_plot = [\n",
    "#     # (\"newN2\", \"newS5\"),\n",
    "#     (\"newN2\", \"newS6\"),\n",
    "#     # (\"newN2\", \"newS7\"),\n",
    "#     (\"oldN2\", \"oldS6\"),\n",
    "#     # (\"sPM57\", \"sPM58\"),\n",
    "#     (\"newerN2\", \"newerS6\"),\n",
    "# ]\n",
    "# plot_titles = list(zip([f\"{libs[0]} vs {libs[1]}\" for libs in lib_comparisons_to_plot], [\"(zoom)\"]*len(lib_comparisons_to_plot)))\n",
    "# \n",
    "# plot_titles = [item for sublist in plot_titles for item in sublist]\n",
    "# \n",
    "# fig = make_subplots(rows=len(lib_comparisons_to_plot), cols=2,\n",
    "#                     column_widths=[0.8, 0.2],\n",
    "#                     shared_yaxes=False,\n",
    "#                     shared_xaxes=True,\n",
    "#                     vertical_spacing=0.1,\n",
    "#                     horizontal_spacing=0.02,\n",
    "#                     subplot_titles=plot_titles)\n",
    "# \n",
    "# for i, (lib1, lib2) in enumerate(lib_comparisons_to_plot):\n",
    "#     plt_df = filter_df.sort_values(f\"{lib1}_v_{lib2}_fishers_p_value\").reset_index()\n",
    "#     p_values = plt_df[f'{lib1}_v_{lib2}_fishers_p_value']\n",
    "#     p_value_ranks = p_values.rank()\n",
    "#     non_1_max_p_value_rank = list(set(plt_df[f'{lib1}_v_{lib2}_fishers_p_value'].rank()))[-2]\n",
    "#     # ^ This value is similar to the number of p-values that I am \"really testing\"\n",
    "#     \n",
    "#     # Overall plots:\n",
    "#     fig.add_trace(go.Scatter(x=p_value_ranks,\n",
    "#                              y=p_values,\n",
    "#                              hovertext=plt_df['gene_name'],\n",
    "#                              mode=\"markers\",\n",
    "#                              marker=dict(color=\"black\"),\n",
    "#                              name=lib2),\n",
    "#                   row=i+1, col=1)\n",
    "#     \n",
    "#     # Add line for a FDR cutoff level using Benjamini-Hochberg method\n",
    "#     fig.add_trace(go.Scatter(y=[(0.05 / non_1_max_p_value_rank), 0.05],\n",
    "#                              x=[1, non_1_max_p_value_rank],\n",
    "#                              mode=\"lines\",\n",
    "#                              marker=dict(color=\"red\"),),\n",
    "#               row=i+1, col=1)\n",
    "#     \n",
    "#     # Zoom plots:\n",
    "#     fig.add_trace(go.Scatter(x=p_value_ranks,\n",
    "#                              y=p_values,\n",
    "#                              hovertext=plt_df['gene_name'],\n",
    "#                              mode=\"markers\",\n",
    "#                              marker=dict(color=\"black\"),\n",
    "#                              name=lib2),\n",
    "#                   row=i+1, col=2)\n",
    "#     fig.add_trace(go.Scatter(y=[(0.05 / non_1_max_p_value_rank), 0.05],\n",
    "#                              x=[1, non_1_max_p_value_rank],\n",
    "#                              mode=\"lines\",\n",
    "#                              marker=dict(color=\"red\"),),\n",
    "#               row=i+1, col=2)\n",
    "#     \n",
    "#     # Name axes:\n",
    "#     fig.update_yaxes(title_text=f\"Fishers P-Value\",\n",
    "#                      row=i+1, col=1)\n",
    "#     \n",
    "#     # Set zoom plot axis limits:\n",
    "#     fig.update_layout(**{f\"yaxis{(i+1)*2}\": dict(range=[0, 0.01]),\n",
    "#                          f\"xaxis{(i+1)*2}\": dict(range=[0, 25])})\n",
    "# fig.update_xaxes(title_text=\"Fishers P-Value Rank<br>N2 vs. smg\",\n",
    "#                  row=4, col=1)\n",
    "# fig.update_xaxes(title_text=\"Fishers P-Value Rank<br>N2 vs. smg\",\n",
    "#                  row=4, col=2)\n",
    "# fig.update_layout(title=f\"Fishers P-value vs Ranking of Fishers P-value\",\n",
    "#                   template=\"plotly_white\",\n",
    "#                   showlegend=False)\n",
    "# fig.show()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "rank_df = filter_df.copy(deep=True).reset_index()\n",
    "rank_df = rank_df.query(\"chr_id != 'MtDNA'\")\n",
    "\n",
    "# This was over-engineered solution, all I really want to check is smg-6 replicable-ity and new smg-5 v new smg-6\n",
    "lib_combos = [(\"oldN2\", \"oldS6\"), (\"newerN2\", \"newerS6\"), (\"newN2\", \"newS6\")]\n",
    "for i, libs1 in enumerate(lib_combos):\n",
    "    libs2 = lib_combos[(i+1) % len(lib_combos)]\n",
    "    # print(libs1, libs2)\n",
    "    \n",
    "    # but I can still use this loop for this step:\n",
    "    lib1, lib2 = libs1\n",
    "    rank_df[f\"{lib1}_v_{lib2}_fishers_p_value_ranked\"] = rank_df[f\"{lib1}_v_{lib2}_fishers_p_value\"].rank()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [
    "lib1, lib2, lib3, lib4 = 'newerN2', 'newerS6', 'oldN2', 'oldS6'\n",
    "\n",
    "minimum_hits_for_both = 100\n",
    "p_value_cutoff_for_both = 1\n",
    "\n",
    "plot_rank_df = rank_df.query(f\"cumulative_unadapted_hits_{lib1}_{lib2} > {minimum_hits_for_both}\")\\\n",
    "                      .query(f\"cumulative_unadapted_hits_{lib3}_{lib4} > {minimum_hits_for_both}\")\n",
    "plot_rank_df = plot_rank_df.query(f'{lib1}_v_{lib2}_fishers_p_value < {p_value_cutoff_for_both}')\\\n",
    "                           .query(f'{lib3}_v_{lib4}_fishers_p_value < {p_value_cutoff_for_both}')\n",
    "\n",
    "fig = px.scatter(plot_rank_df,\n",
    "                 x=f'{lib1}_v_{lib2}_fishers_p_value_ranked',\n",
    "                 y=f'{lib3}_v_{lib4}_fishers_p_value_ranked',\n",
    "                 hover_name='gene_name',\n",
    "                 # hover_data=[''],\n",
    "                 # log_x=True, log_y=True,\n",
    "                 trendline='ols',\n",
    "                 title=f\"Comparison of ranked p_values for {lib1}_v_{lib2} and {lib3}_v_{lib4}\"\n",
    "                 )\n",
    "fig.show()\n",
    "\n",
    "fig = px.scatter(plot_rank_df,\n",
    "                 x=f'{lib1}_v_{lib2}_fishers_p_value',\n",
    "                 y=f'{lib3}_v_{lib4}_fishers_p_value',\n",
    "                 hover_name='gene_name',\n",
    "                 # hover_data=[''],\n",
    "                 # log_x=True, log_y=True,\n",
    "                 trendline='ols',\n",
    "                 title=f\"Comparison of p_values for {lib1}_v_{lib2} and {lib3}_v_{lib4}\"\n",
    "                 )\n",
    "fig.show()"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "markdown",
   "source": [
    "Plotting comparison of p-values has helped me appreciate that the majority of genes failed to be replicated between the libraries. Some genes of interest remain highly ranked in both of the library sets:\n",
    "* F19B2.5   This is a fairly interesting little gene that only has 1 isoform. It obviously has some kind of NMD dependent cleavage or degradation event happening (the coverage plots in wildtype are slopes while they're pretty square in nmd knockouts)"
   ],
   "metadata": {
    "collapsed": false
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
