{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import sys, os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Supplemental_File_2.csv\n",
      "  Unnamed: 0  Unnamed: 1 Unnamed: 2 trans-splice site 1 Unnamed: 4  \\\n",
      "0      gene   chromosome     strand            position  SL1 reads   \n",
      "1    F45E6.3           X          +            12484975         18   \n",
      "2    F45E6.2           X          +            12480870        106   \n",
      "3    Y47A7.2           V          -             5198152        263   \n",
      "4    Y47A7.1           V          +             5204291         52   \n",
      "\n",
      "                            Unnamed: 5    Unnamed: 6 trans-splice site 2  \\\n",
      "0  SL2 reads (SL2 only no SL2 variant)  Sl2 variant             position   \n",
      "1                                   20            21                 NaN   \n",
      "2                                    1             4                 NaN   \n",
      "3                                    0             1                 NaN   \n",
      "4                                    1             0                 NaN   \n",
      "\n",
      "  Unnamed: 8                           Unnamed: 9  ...  \\\n",
      "0  SL1 reads  SL2 reads (SL2 only no SL2 variant)  ...   \n",
      "1        NaN                                  NaN  ...   \n",
      "2        NaN                                  NaN  ...   \n",
      "3        NaN                                  NaN  ...   \n",
      "4        NaN                                  NaN  ...   \n",
      "\n",
      "                           Unnamed: 41   Unnamed: 42 trans-splice site 11  \\\n",
      "0  SL2 reads (SL2 only no SL2 variant)  Sl2 variant              position   \n",
      "1                                  NaN           NaN                  NaN   \n",
      "2                                  NaN           NaN                  NaN   \n",
      "3                                  NaN           NaN                  NaN   \n",
      "4                                  NaN           NaN                  NaN   \n",
      "\n",
      "  Unnamed: 44                          Unnamed: 45   Unnamed: 46  \\\n",
      "0   SL1 reads  SL2 reads (SL2 only no SL2 variant)  Sl2 variant    \n",
      "1         NaN                                  NaN           NaN   \n",
      "2         NaN                                  NaN           NaN   \n",
      "3         NaN                                  NaN           NaN   \n",
      "4         NaN                                  NaN           NaN   \n",
      "\n",
      "  trans-splice site 12 Unnamed: 48                          Unnamed: 49  \\\n",
      "0             position   SL1 reads  SL2 reads (SL2 only no SL2 variant)   \n",
      "1                  NaN         NaN                                  NaN   \n",
      "2                  NaN         NaN                                  NaN   \n",
      "3                  NaN         NaN                                  NaN   \n",
      "4                  NaN         NaN                                  NaN   \n",
      "\n",
      "    Unnamed: 50  \n",
      "0  Sl2 variant   \n",
      "1           NaN  \n",
      "2           NaN  \n",
      "3           NaN  \n",
      "4           NaN  \n",
      "\n",
      "[5 rows x 51 columns]\n",
      "Supplemental_file_4.csv\n",
      "                          Genes > 0.05 average dcpm   \\\n",
      "0                                      With Go terms   \n",
      "1                                                NaN   \n",
      "2                                            Go term   \n",
      "3  GO:0009790,%biological_process%developmental p...   \n",
      "4  GO:0032502,%biological_process%developmental p...   \n",
      "\n",
      "                                                  n=                  9975  \\\n",
      "0                                                 n=                  6507   \n",
      "1                                                NaN                   NaN   \n",
      "2                 trans-spliced genes with this term  trans-spliced genes    \n",
      "3  deps-1;nasp-2;npp-11;dhc-1;lsm-5;b0285.1;k10d2...                  2119   \n",
      "4  deps-1;sma-9;nasp-2;npp-11;dhc-1;lsm-5;lin-2;b...                  2741   \n",
      "\n",
      "                       3917 Unnamed: 4  \n",
      "0                      1842        NaN  \n",
      "1                       NaN        NaN  \n",
      "2  non-trans-spliced genes     p-value  \n",
      "3                       260        0.0  \n",
      "4                       411        0.0  \n",
      "Supplemental_File_3.csv\n",
      "  Min percent SL2 to be counted as downstream in an operon     0.1  \\\n",
      "0                                                NaN        number   \n",
      "1                            total predicted operons          1429   \n",
      "2                            total annotated operons          1267   \n",
      "3                                total exact matches           913   \n",
      "4                                                NaN           NaN   \n",
      "\n",
      "             Unnamed: 2                    Unnamed: 3  \n",
      "0  percent of annotated  percent of predicted operons  \n",
      "1                   NaN                           NaN  \n",
      "2                   NaN                           NaN  \n",
      "3                0.7206                      0.638908  \n",
      "4                   NaN                           NaN  \n",
      "Supplemental_File_1.csv\n",
      "  Chromosome strand  trans-spliced position  percent SL2  \\\n",
      "0          I      +                   11618     0.001197   \n",
      "1          I      +                   47463     0.004132   \n",
      "2          I      +                   49921     0.907895   \n",
      "3          I      +                   71847     0.000000   \n",
      "4          I      +                   72511     0.000000   \n",
      "\n",
      "   binomial exact p-value for the percent SL2  number of reads  \\\n",
      "0                                0.000000e+00             5014   \n",
      "1                                0.000000e+00              484   \n",
      "2                                3.808720e-27              152   \n",
      "3                                0.000000e+00              270   \n",
      "4                                9.860760e-32              103   \n",
      "\n",
      "   number of SL1 reads  number of total SL2 reads (SL2+SL2 variant reads)  \\\n",
      "0                 5008                                                  6   \n",
      "1                  482                                                  2   \n",
      "2                   14                                                138   \n",
      "3                  270                                                  0   \n",
      "4                  103                                                  0   \n",
      "\n",
      "   number of SL2 variant reads  non-trans-spliced reads at this position  ...  \\\n",
      "0                            6                                         0  ...   \n",
      "1                            2                                         0  ...   \n",
      "2                           98                                         4  ...   \n",
      "3                            0                                         6  ...   \n",
      "4                            0                                         8  ...   \n",
      "\n",
      "   gene name gene rank (within operon)  position of first AUG  \\\n",
      "0   Y74C9A.2                         0                  11640   \n",
      "1  Y48G1C.12                         1                  47471   \n",
      "2   Y48G1C.4                         2                  50450   \n",
      "3   Y48G1C.2                         1                  71987   \n",
      "4   Y48G1C.2                         1                  72530   \n",
      "\n",
      "   average exp of the gene (dcpm) distance to upstream gene 3' end  \\\n",
      "0                         9.13074                                0   \n",
      "1                        0.498897                             2786   \n",
      "2                        0.354835                               64   \n",
      "3                         1.03737                            17422   \n",
      "4                         1.03737                            18086   \n",
      "\n",
      "   position of the upstream gene 3' end  \\\n",
      "0                                     0   \n",
      "1                                 44677   \n",
      "2                                 49857   \n",
      "3                                 54425   \n",
      "4                                 54425   \n",
      "\n",
      "   length of 5' UTR (trans-splice site to first AUG)  \\\n",
      "0                                                 22   \n",
      "1                                                  8   \n",
      "2                                                529   \n",
      "3                                                140   \n",
      "4                                                 19   \n",
      "\n",
      "   position of the closest annotated 5' end  \\\n",
      "0                                     11618   \n",
      "1                                     47472   \n",
      "2                                     49919   \n",
      "3                                     71858   \n",
      "4                                     72511   \n",
      "\n",
      "   distance from the closest annotated 5' end to trans-splice site  \\\n",
      "0                                                  0                 \n",
      "1                                                  9                 \n",
      "2                                                 -2                 \n",
      "3                                                 11                 \n",
      "4                                                  0                 \n",
      "\n",
      "   distance from the closest annotated 5' end to first AUG downstream of trans-splice site  \n",
      "0                                                -22                                        \n",
      "1                                                  1                                        \n",
      "2                                               -531                                        \n",
      "3                                               -129                                        \n",
      "4                                                -19                                        \n",
      "\n",
      "[5 rows x 21 columns]\n"
     ]
    }
   ],
   "source": [
    "for file_name in os.listdir():\n",
    "    if file_name.endswith(\".xls\"):\n",
    "        temp_df = pd.read_excel(file_name)\n",
    "        save_name = file_name[:-3] + \"csv\"\n",
    "        temp_df.to_csv(save_name, index=False)\n",
    "        print(save_name)\n",
    "        print(temp_df.head())"
   ],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "outputs": [],
   "source": [],
   "metadata": {
    "collapsed": false,
    "pycharm": {
     "name": "#%%\n"
    }
   }
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}